#[instruction_set]
//~^ ERROR the `#[instruction_set]` attribute is an experimental feature [E0658]
//~| ERROR malformed `instruction_set` attribute input
//~| ERROR must specify an instruction set [E0778]
fn main() {
}
